p2panda-rs/src/entry/validate.rs

```diff
@@ -0,0 +1,301 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+//! Collection of low-level validation methods for entries.
+//!
+//! You will not find methods here to check the encoding of Bamboo entries, as this is handled
+//! inside the external bamboo-rs crate.
+use crate::entry::error::ValidateEntryError;
+use crate::entry::{EncodedEntry, Entry, Signature};
+use crate::hash::Hash;
+use crate::identity::{Author, KeyPair};
+use crate::operation::EncodedOperation;
+
+/// Checks if backlink- and skiplink are correctly set for the given sequence number (#E3).
+///
+/// First entries do not contain any links. Every other entry has to contain a back- and skiplink
+/// unless they are equal, in which case the skiplink must be omitted.
+pub fn validate_links(entry: &Entry) -> Result<(), ValidateEntryError> {
+    match (
+        entry.seq_num().is_first(),
+        entry.backlink().is_some(),
+        entry.skiplink().is_some(),
+        entry.is_skiplink_required(),
+    ) {
+        (true, false, false, false) => Ok(()),
+        (false, true, false, false) => Ok(()),
+        (false, true, true, _) => Ok(()),
+        (_, _, _, _) => Err(ValidateEntryError::InvalidLinks),
+    }?;
+
+    if entry.is_skiplink_required() && entry.backlink() == entry.skiplink() {
+        return Err(ValidateEntryError::BacklinkAndSkiplinkIdentical);
+    }
+
+    Ok(())
+}
+
+/// Checks if entry is correctly placed in its log (#E4).
+///
+/// The following validation steps are applied:
+///
+/// 1. Are the claimed backlink and skiplink entries part of the same log?
+/// 2. Are the claimed backlinks and skiplinks published by the same key?
+/// 3. Are the claimed backlink and skiplink hashes matching with what is in the log?
+///
+/// This method requires knowledge about other entries. Use this together with your storage
+/// provider implementation.
+pub fn validate_log_integrity(
+    entry: &Entry,
+    skiplink_entry: Option<&Entry>,
+    skiplink_entry_hash: Option<&Hash>,
+    backlink_entry: Option<&Entry>,
+    backlink_entry_hash: Option<&Hash>,
+) -> Result<(), ValidateEntryError> {
+    // First make sure we didn't pass wrong arguments
+    if backlink_entry.is_some() != backlink_entry_hash.is_some() {
+        panic!("Both encoded and decoded backlink entry must be the same")
```

Comment:
I'm fine with either, this post I read today and actually I think still panicking would fit: https://blog.burntsushi.net/unwrap/ as it would be a buggy implementation but using the tuples is also nice!